pkg/endpoint/policy.go

```diff
@@ -110,25 +112,62 @@ func getSecurityIdentities(labelsMap *LabelsMap, selector *api.EndpointSelector)
 	return identities
 }
 
-func (e *Endpoint) removeOldFilter(labelsMap *LabelsMap, filter *policy.L4Filter) {
+// removeOldFilter removes the old l4 filter from the endpoint.
+// Returns a map that represents all policies that were attempted to be removed;
+// it maps to whether they were removed successfully (true or false)
+// It also returns the number of errors that occurred while when removing the
+// policy.
+func (e *Endpoint) removeOldFilter(owner Owner, labelsMap *LabelsMap,
+	filter *policy.L4Filter) policy.RuleContexts {
+
+	fromEndpointsSrcIDs := policy.RuleContexts{}
 	port := uint16(filter.Port)
 	proto := uint8(filter.U8Proto)
 
 	for _, sel := range filter.FromEndpoints {
 		for _, id := range getSecurityIdentities(labelsMap, &sel) {
 			srcID := id.Uint32()
+			ruleCtx := policy.RuleContext{
+				SecID:          id,
+				Port:           byteorder.HostToNetwork(port).(uint16),
+				Proto:          proto,
+				L7RedirectPort: byteorder.HostToNetwork(uint16(filter.L7RedirectPort)).(uint16),
+				IsRedirect:     filter.IsRedirect(),
+			}
 			if err := e.PolicyMap.DeleteL4(srcID, port, proto); err != nil {
 				// This happens when the policy would add
 				// multiple copies of the same L4 policy. Only
 				// one of them is actually added, but we'll
 				// still try to remove it multiple times.
 				e.getLogger().WithError(err).WithField(logfields.L4PolicyID, srcID).Debug("Delete old l4 policy failed")
+				// Set with false only if the key was not
+				// previously set nor the value was set
+				// with true. Since we can have multiple
+				// copies of the same L4 policy, a single
+				// successful DeleteL4 means the policy
+				// was actually removed.
+				v, ok := fromEndpointsSrcIDs[ruleCtx]
```

Comment:
This is relatively fragile, it only works because all members of `RuleContext` are comparable with `==`. Please add a comment to `RuleContext struct` that no field can be introduced which is not comparable with `==`

Reply:
Thanks for pointing this out. Golang will now if a key is not comparable so it won't compile it. I still add the following comment:

`// Don't use pointers here since this structure is used as key on maps.`